engine/access/rpc/backend/connection_factory.go

```diff
@@ -74,24 +88,63 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 
 func (cf *ConnectionFactoryImpl) retrieveConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, error) {
 	var conn *grpc.ClientConn
+	var mutex *sync.Mutex
 	if res, ok := cf.ConnectionsCache.Get(grpcAddress); ok {
-		conn = res.(*grpc.ClientConn)
+		conn = res.(ConnectionCacheStore).ClientConn
+		mutex = res.(ConnectionCacheStore).mutex
+
+		// we lock this mutex to prevent a scenario where the connection is not good, which will result in
+		// re-establishing the connection for this address. if the mutex is not locked, we may attempt to re-establish
+		// the connection multiple times which would result in cache thrashing.
+		mutex.Lock()
+		defer mutex.Unlock()
 		if cf.AccessMetrics != nil {
 			cf.AccessMetrics.ConnectionFromPoolRetrieved()
 		}
 	}
 	if conn == nil || conn.GetState() != connectivity.Ready {
-		// This line ensures that when a connection is renewed, the previously cached connection is evicted and closed
-		cf.ConnectionsCache.Remove(grpcAddress)
+		// this lock prevents a memory leak where a race condition may occur if 2 requests to a new connection at the
+		// same address occur. the second add would overwrite the first without closing the connection
+		cf.lock.Lock()
+		defer cf.lock.Unlock()
+		// Check if connection was created/refreshed by another thread
+		if res, ok := cf.ConnectionsCache.Get(grpcAddress); ok {
+			conn = res.(ConnectionCacheStore).ClientConn
+			if conn.GetState() == connectivity.Ready {
```

Comment:
are you sure `Ready` is the right measure here? it looks like anything but `Shutdown` could be ok